ndcube/ndcube.py

```diff
@@ -345,21 +372,51 @@ def extra_coords(self):
                     "value": self._extra_coords_wcs_axis[key]["value"]}
         return result
 
-    def crop(self, *intervals, wcs=None):
+    def crop(self, lower_corner, upper_corner, wcs=None):
         # The docstring is defined in NDCubeBase
-        return self._crop(*intervals, wcs=wcs, crop_by_values=False)
+        if len(lower_corner) != len(upper_corner):
+            raise ValueError("lower_corner must have same length as upper_corner, "
+                             f"lower_corner: {lower_corner}; upper_corner: {upper_corner}")
+        return self._crop(lower_corner, upper_corner, wcs, False)
 
-    def crop_by_values(self, *intervals, wcs=None):
+    def crop_by_values(self, lower_corner, upper_corner, units=None, wcs=None):
         # The docstring is defined in NDCubeBase
-        return self._crop(*intervals, wcs=wcs, crop_by_values=True)
-
-    def _crop(self, *intervals, wcs=None, crop_by_values=False):
+        # Sanitize inputs.
+        n_world_dims = len(lower_corner)
+        if len(upper_corner) != n_world_dims:
+            raise ValueError("lower_corner must have same length as upper_corner")
+        if units is None:
+            units = [None] * n_world_dims
+        elif len(units) != n_world_dims:
+            raise ValueError(
+                "units must be None or have same length as lower_corner and upper_corner.")
+        # Convert float inputs to quantities using units.
```

Comment:
> So we are allowing mixed mode here? Some args can be quantity and some float?

Yes.  I don't see why not.  If we didn't allow it we should still check to make sure all types are the same.  Otherwise unexpected behaviour could potentially occur.  And in that case we would have to introduce some code complexity here anyway.

> Would this logic be simpler if we just said "ok so units isn't None, so all your inputs get converted to given units".

I don't think so.  Because you could have
```python
lower_corner = [None]
upper_corner = [2]
units = [u.m]
```
In this case even though units is set, `lower_corner[0]` remains `None`.